### ISSUE TITLE
defining class attribute after use doesn't throw an error 

### ISSUE THREAD

**Issue by Goradii** [microsoft/pyright#7632]
<h3> Environment data </h3>

-   Language Server version: 2024.4.1
-   OS and version: linux x64
-   Python version 3.11
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: basic

<h3> Code Snippet </h3>

```python

class Test:
    def __init__(self):
        pass

    def test(self):
        if self._test_var: # no error, but cause an error in runtime
            self._test_var = False

```

<h3> Expected behavior </h3>

Cannot access member "_test_var" for type "Test*"
  Member "_test_var" is unknownPylance[reportAttributeAccessIssue]

or 

Cannot access member "_test_var" for type "Test*"
  Member "_test_var" doesn't defined it __init__

<h3> Actual behavior </h3>

no error

**Comment by erictraut** [comment]
Pyright is working as designed here. This isn't a bug or a missing feature.

While this does generate an exception at runtime, this class of bug cannot be caught through static analysis — at least not the type of static analysis provided by a type checker like pyright. The problem is that instance variables can be assigned (or deleted) within multiple execution contexts (in the class body, in various methods or functions), etc. They can be set (or deleted) in subclasses defined in other modules. They can also be assigned in subclasses via mix-ins using multiple inheritance. 

```python
class Test:
    def __init__(self):
        pass

    def test(self):
        if self._test_var: # no error, but cause an error in runtime
            self._test_var = False

    def test2(self):
        self._test_var = 1

t1 = Test()
t1.test2()
t1.test() # No error

t2 = Test()
t2._test_var = 1
t2.test() # No error

class TestChild(Test):
    def __init__(self):
        super().__init__()
        self._test_var = 1

t3 = TestChild()
t3.test() # No error
```

Pyright does offer a diagnostic check called `reportUninitializedInstanceVariable` that verifies that all instance variables are initialized in either the class body or in the `__init__` method. This rule requires a very specific coding style that most Python programmers find a bit too constraining. But if you're willing to adopt this programming style, you can catch bugs like this one statically.

To enable this rule in a single module, add `# pyright: reportUninitializedInstanceVariable=true` to the top of the file. You can also add it for your entire project using a pyrightconfig.json or pyproject.toml file or through your language server settings. For more details, refer to [this documentation](https://microsoft.github.io/pyright/#/configuration?id=main-configuration-options).
